mds: record and forward client side metrics to manager

vshankar opened the pull request:
This is a WIP and has basic messaging boilerplates for an MDS to forward metrics (client metrics, etc..) to rank 0. Follow up PRs would stitch all this together with clients periodically sending its metrics to its session MDS with this MDS forwarding it to rank 0 (which would aggregate all session metadata and update the mgr).

batrick commented:
`Partially-fixes: http://tracker.ceph.com/issues/36253`

batrick commented:
@vshankar please rebase this when you get a chance.

vshankar commented:
> @vshankar please rebase this when you get a chance.

yep, on it w/ the changes to forwarding metrics to the manager.

vshankar commented:
Still keeping this as WIP -- I should probably include a detailed implementation writeup here (or in the commit message). I'll do that on priority -- but this should be ok for a round of review if someone wants to delve into the details right now!

vshankar commented:
> please rebase

I'll get to this next week.

vshankar commented:
@batrick -- rebased + updated. please take a look.

batrick commented on src/mds/ServiceDaemon.h:
```diff
@@ -0,0 +1,37 @@
+
+class MgrClient;
+
+class ServiceDaemon {
```
Let's also make this subclass `Dispatcher` like `Beacon`. I would like to see it support fast dispatch of metrics from clients and from other MDSs (if rank 0). It'd probably be best to split this into two classes: `MetricGatherer` (for client metric aggregation for transmission to rank 0) and `MetricAggregator` (rank 0 only) which handles aggregate client messages from its own `MetricGatherer` (special case) and from other ranks' `MetricGatherer`s. (Keep in mind that actual aggregation of a client's metrics received from all other ranks takes place in `MetricAggregator`; e.g. adding all cap hits.)

For fast dispatch, I think a spinlock would work for inserting metric messages into a metrics queue for the `MetricGatherer`/`MetricAggregator` thread to digest. What do you think?

batrick replied:
(If you're a little confused by my previous comment, I'm going through your PR by commit so I'm only now seeing `MetricsHandler`.)

I see now `MetricsHandler`, that's similar to the `MetricGatherer` I suggested above. I'd still like to see that as a separate dispatcher. I don't want these metrics message unnecessarily stuck in the MDS message queue. So it'd be broken out from `Server` and be part of `MDSRank`.

vshankar replied:
> (If you're a little confused by my previous comment, I'm going through your PR by commit so I'm only now seeing `MetricsHandler`.)

yeh -- I kind of split that out already.

> 
> I see now `MetricsHandler`, that's similar to the `MetricGatherer` I suggested above. I'd still like to see that as a separate dispatcher. I don't want these metrics message unnecessarily stuck in the MDS message queue. So it'd be broken out from `Server` and be part of `MDSRank`.

Makes sense to have that as a separate dispatcher. I'll do the required changes.

batrick commented on src/mds/ServiceDaemon.cc:
```diff
@@ -0,0 +1,99 @@
+int ServiceDaemon::init() {
+  dout(20) << dendl;
+
+  std::lock_guard<ceph::mutex> locker(lock);
```
`std::lock_guard` is considered obsolete. Prefer instead `std::scoped_lock`. You should also be able to get rid of the `<ceph::mutex>` due to template deduction.

vshankar replied:
ack

batrick commented on src/mds/MDSPerfMetricTypes.h:
```diff
@@ -0,0 +1,116 @@
+  RankMetrics(mds_rank_t rank) : rank(rank) {
+  }
+
+  mds_rank_t rank = MDS_RANK_NONE;
```
Let's also add an sequence number (epoch) here so that rank 0 can determine if another rank is behind. If there is no update, a non-zero rank can still send an update with the new epoch. Likewise, rank 0 needs to send out a heartbeat with a new epoch of its own. `RankMetrics` should include the new sequence number and the last epoch it saw from rank 0.

vshankar commented on src/mds/metrics/client/Types.h:
```diff
@@ -0,0 +1,105 @@
+typedef boost::variant<CapInfoPayload,
+                       OSDCCacheInfoPayload,
+                       ReadWriteIOPSPayload,
+                       UnknownPayload> ClientMetricPayload;
```
Mostly, yes. But splitting it out will give the flexibility to send only what required (changed, maybe)?

vshankar commented on src/mgr/MDSPerfMetricCollector.cc:
```diff
@@ -0,0 +1,22 @@
+#undef dout_prefix
+#define dout_prefix *_dout << "mgr.mds_perf_metric_collector " << __func__ << " "
+
+void MDSPerfMetricCollector::process_reports(const ClientRankMetrics &report) {
```
ack

vshankar commented on src/mds/metrics/client/Types.cc:
```diff
@@ -0,0 +1,176 @@
+
+// vairant : cap info
+void CapInfoPayload::encode(bufferlist &bl) const {
+  using ceph::encode;
```
yep -- I already have version additions locally.

batrick commented:
@vshankar what's the status of this PR? Is it blocking on my review?

vshankar commented:
> @vshankar what's the status of this PR? Is it blocking on my review?

kind of -- although I am waiting for more comments on PR #29214 so I can rebase.

If you want you can go ahead with a round of review. I addressed most of your comments.

vshankar commented:
retest this please

vshankar commented:
@batrick please take a look. I'll write a list of changes in this patchset so it's easier for review.

vshankar commented on src/mds/metrics/client/Types.h:
```diff
@@ -0,0 +1,105 @@
+};
+std::ostream &operator<<(std::ostream &os, const ClientMetricType &type);
+
+struct CapInfoPayload {
```
I think we should just forward raw counters up to the ceph-mgr. That way, its upto the manager module to display the metric as mere raw perfs, %ages and other aggregation mechanisms (depending on manager plugin command-line options?).

I have kept this as it is w/ the latest pr update.

vshankar commented:
changes:

- `MetricAggregator` makes use of MDS query and performance types to collect metrics. Maintaining a map of user query to metrics helps in selecting only those performance descriptors that are requested by the user. Moreover, this is customizable by providing a regex to include/exclude performance descriptors. eg. an operator can run a query (via ceph mgr) such as:

        > ceph fs perf stats --mds_rank=1 --client=10.0.0.1

to only show metrics from mds rank 1 for a selected client IP address.

- `MDSPinger` class for sending ping messages to all active ranks to identify which ranks are lagging. This is used by `MetricAggregator` for sending ping message, "acking" to pong responses and checking if an MDS rank is lagging. Metrics from laggy ranks are marked as "delayed" in metrics reports that are sent to ceph-mgr. `MetricAggregator` also gets notified on mdsmap changes for culling out metric entries for failed ranks and initializing fresh metrics for new active ranks.

- incorporated comments

vshankar commented on src/mds/metrics/client/Types.cc:
```diff
@@ -0,0 +1,176 @@
+
+// variant : unknown
+void UnknownPayload::encode(bufferlist &bl) const {
+  ceph_abort();
```
I haven't fixed this with the latest update -- but, is this possible w/ the upgrade steps of setting max_mds=1 and upgrading the mds cluster first?

batrick replied:
Late reply sorry: no. I mean we could have e.g. a P client talking to an O cluster. If it sends an unknown payload, the MDS shouldn't kill itself.

vshankar replied:
I changed this to a noop w/ the last update.

batrick commented on src/mds/MDSPinger.cc:
```diff
@@ -0,0 +1,109 @@
+    dout(20) << ": init ping pong state for rank=" << rank << dendl;
+    it = ping_state_by_rank.emplace(rank, PingState()).first;
+    it->second.last_acked_time = clock::now();
+  }
```
this code can be simplified:

````
auto [it, inserted] = ping_state_by_rank.emplace(rank, {});
if (inserted) {
   dout(20) << ": init ping pong state for rank=" << rank << dendl;
}
````

(you already initialize `last_acked_time` in the header)

batrick commented on src/mds/MDSRank.cc:
```diff
@@ -1210,6 +1212,11 @@ bool MDSRank::handle_deferrable_message(const cref_t<Message> &m)
 
+    case MSG_MDS_PING:
+      ALLOW_MESSAGES_FROM(CEPH_ENTITY_TYPE_MDS);
+      server->dispatch(m);
```
Wouldn't we also want to fast dispatch these?

batrick replied:
I'm confused @vshankar, I'm looking at the current code and this is still done through regular dispatch? Did you forget to commit some changes?

batrick commented:
also needs rebase

vshankar commented on src/mds/Server.cc:
```diff
@@ -267,9 +268,15 @@ void Server::dispatch(const cref_t<Message> &m)
     handle_client_reclaim(ref_cast<MClientReclaim>(m));
     return;
+  case CEPH_MSG_CLIENT_METRICS:
+    handle_client_metrics(ref_cast<MClientMetrics>(m));
```
yeh -- I kind of spent time converting `MetricAggregator` into a separate dispatcher and completely ignored this. I'll do the required changes here too.

vshankar commented on src/mds/MDSDaemon.cc:
```diff
@@ -877,6 +878,14 @@ void MDSDaemon::handle_mds_map(const cref_t<MMDSMap> &m)
       mds_rank->init();
+
+      // metric aggregation is solely done by rank 0
+      if (mds_rank->get_nodeid() == (mds_rank_t)0) {
```
ah! yes -- didn't fully understand how this would work with standby+replay. I'll fix this and add a test case too.

vshankar commented on src/mds/MetricAggregator.cc:
```diff
@@ -0,0 +1,356 @@
+           << metrics << dendl;
+
+  auto p = clients_by_rank.find(rank);
+  ceph_assert(p != clients_by_rank.end());
```
ack -- good to know this.

vshankar commented:
@batrick -- apologies for the delay -- have been in and out of other things.

pushing an update today.

batrick commented:
> I'm confused @vshankar, I'm looking at the current code and this is still done through regular dispatch? Did you forget to commit some changes?

Same question for my other "resolved" comments.

vshankar commented:
> Same question for my other "resolved" comments.

I marked the comments as resolved, but "git push" was failing yesterday.

vshankar commented:
> I marked the comments as resolved, but "git push" was failing yesterday.

will push tomorrow when I have access...

vshankar commented:
@batrick updated -- please take a look.

vshankar commented:
changes in the latest patchset:

* standby+replay fixes -- earlier standby+replay MDSs were sending perf metric to active rank 0
* `MetricsHandler` and `MetricAggregator` have their own dispatch queues thereby not contending for `mds_lock`
* all messages (mds -> mds, client -> mds) are versioned
* code improvements by Patrick
* tests cases are in pr #29951 (needs manager module for fetching metrics)

vshankar commented on src/common/CMakeLists.txt:
```diff
@@ -97,7 +97,8 @@ set(common_srcs
   version.cc
-  xattr.c)
+  xattr.c
+  ${CMAKE_SOURCE_DIR}/src/mds/metrics/client/Types.cc)
```
I've not incorporated this yet.

tchaikov replied:
@batrick sorry for the latency. i am actually surprised that we are putting mds specific stuff into ceph-common. how is it used by other components? can we include it in `mds_srcs` , and link against `mds` instead ?

batrick replied:
>  i am actually surprised that we are putting mds specific stuff into ceph-common

That's just where it is pending the result of this discussion :)

> how is it used by other components?

client and MDS

> can we include it in mds_srcs , and link against mds instead ?

Perhaps just move it to a header file in `src/include` and get rid of `Types.cc`? I don't like the idea of the client linking to code in the MDS but we kinda already do this for `src/tools/cephfs`. So it wouldn't be too big a departure from what we already do.

tchaikov replied:
agreed. we are not doing great in this area.. probably we can avoid making this worse by going  further in this direction?

even though we have libceph-common.so, libceph-common.a is still statically linked into quite a few tests and ceph tools. i really want to avoid putting more objects into that bucket. for couple reasons, it takes more than 1 hour to upload the debuginfo now. and probably more important is that it just does not look/sound right ...

batrick replied:
Does moving the entire thing Types.cc/Types.h into a header in /src/include  sound good?

batrick replied:
Actually, the header could go in `src/mds`, we already link to some MDS headers in the client.

tchaikov replied:
> Does moving the entire thing Types.cc/Types.h into a header in /src/include sound good?

sounds good to me.

tchaikov commented on src/mds/MDSPerfMetricTypes.h:
```diff
@@ -0,0 +1,189 @@
+#ifndef CEPH_MDS_PERF_METRIC_TYPES_H
+#define CEPH_MDS_PERF_METRIC_TYPES_H
+
+#include <iostream>
```
ostream would suffice?

tchaikov commented on src/mds/MDSPerfMetricTypes.h:
```diff
@@ -0,0 +1,189 @@
+#include "include/utime.h"
+#include "mdstypes.h"
+
+enum UpdateType {
```
```suggestion
enum UpdateType : uint32_t {
```

tchaikov commented on src/mds/MDSPerfMetricTypes.h:
```diff
@@ -0,0 +1,189 @@
+    using ceph::encode;
+
+    ENCODE_START(1, 1, bl);
+    encode(static_cast<uint32_t>(update_type), bl);
```
no need to cast the enum to the underlying type if it has an underlying type like
```c++
enum UpdateType : uint32_t
```

tchaikov commented on src/mds/MDSPerfMetricTypes.h:
```diff
@@ -0,0 +1,189 @@
+  void dump(Formatter *f) const {
+    f->dump_unsigned("seq", seq);
+    f->dump_int("rank", rank);
+    for (auto &p : client_metrics_map) {
```
```suggestion
    for (auto &[client, metrics] : client_metrics_map) {
```

tchaikov commented on src/mds/MDSPerfMetricTypes.h:
```diff
@@ -0,0 +1,189 @@
+    : seq(seq), rank(rank) {
+  }
+
+  void encode(bufferlist &bl, uint64_t features) const {
```
nit, could use `DENC` macro instead. like
```c++
DENC(MetricsMessage, v, p) {
  DENC_START(1, 1, p);
  denc(v.seq, p);
  /// ...
  DENC_FINISH(p);
}
```

vshankar replied:
`client_metrics_map` field in this class uses the old-style encoding. Is it ok for this class to use new-style encoding? I tried this in the past but couldn't get it working.

tchaikov replied:
i don't see a reason why we cannot start using new-style encoding here. 

> but couldn't get it working.

could you elaborate a little bit?

vshankar replied:
> i don't see a reason why we cannot start using new-style encoding here.
> 
> > but couldn't get it working.
> 
> could you elaborate a little bit?

well, it was a while back -- do not remember the specifics, but somewhat relating to `denc` macro unable to encode/decode member variable using old-style encoders.

I'll try again.

vshankar replied:
@tchaikov -- need a bit of help in converting to new-style encoding defs:

For this class, I changed it as per:

```
+  DENC_FEATURED(metrics_message_t, v, p, f) {
+    DENC_START(1, 1, p);
+    denc(v.seq, p);
+    denc(v.rank, p);
+    denc(v.client_metrics_map, p, f);
+    DENC_FINISH(p);
+  }

+WRITE_CLASS_DENC_FEATURED(metrics_message_t)
```
(_FEATURED version is required since `entity_inst_t` need to be encoded).

With this change I get the following compilation err:

```
/work/ceph/src/mds/MDSPerfMetricTypes.h: In instantiation of ‘std::enable_if_t<(is_same_v<T, metrics_message_t> || is_same_v<T, const metrics_message_t>)> _denc_friend(T&, P&, uint64_t) [with T = const metrics_message_t; P = long unsigned int; std::enable_if_t<(is_same_v<T, metrics_message_t> || is_same_v<T, const metrics_message_t>)> = void; uint64_t = long unsigned int]’:
/work/ceph/src/mds/MDSPerfMetricTypes.h:132:3:   required from here
/work/ceph/src/mds/MDSPerfMetricTypes.h:136:9: error: no matching function for call to ‘denc(const std::map<entity_inst_t, Metrics>&, long unsigned int&, uint64_t&)’
     denc(v.client_metrics_map, p, f);
     ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I tried looking at include/denc.h but couldn't get a deep understanding of the macros.  So, for testing, I change the std::map to be `<uint32_t, Metrics>` and that compiles fine. Seems to me that some additional handling needs to be done for `entity_inst_t`?

vshankar replied:
@batrick ^^

batrick replied:
Hmm, you're in the weeds aren't you. I'm not sure. Let's wait and see if @tchaikov knows off-hand what's wrong.

tchaikov replied:
@vshankar sorry, i was wrong -- the new-style denc requires the member variables to be encoded are also properly annotated with `denc_traits<>` definitions. and to switch `entity_addr_t` to denc machinery is not trivial. so i am resolving this conversation.

tchaikov commented on src/mds/MDSPerfMetricTypes.h:
```diff
@@ -0,0 +1,189 @@
+
+// Initial Sequence Number (ISN) of the first ping message sent
+// by rank 0 to other active ranks (incuding itself).
+#define MDS_PINGER_ISN 1
```
is it used anywhere else than `MDSPinger`? if not, why not define it there as a `static constexpr` member variable?

vshankar replied:
ok -- yes, I might have had a plan to use this in `MetricsHandler`, but didn't find a need. Will move  this to `MDSPinger` class.

tchaikov commented on src/mds/MDSPerfMetricTypes.h:
```diff
@@ -0,0 +1,189 @@
+// by rank 0 to other active ranks (incuding itself).
+#define MDS_PINGER_ISN 1
+
+struct MetricsMessage {
```
if it's a plain struct for data, might want to name it like `metrics_message_t`.

batrick replied:
+1 MetricsMessage makes you think it comes from `src/messages`.

vshankar replied:
ACK

tchaikov commented on src/mds/MetricAggregator.cc:
```diff
@@ -0,0 +1,345 @@
+    if (m->get_connection()->get_peer_type() == CEPH_ENTITY_TYPE_MDS) {
+      handle_mds_metrics(ref_cast<MMDSMetrics>(m));
+    }
+
```
please remove empty lines like this.

batrick commented:
needs rebase too

vshankar commented:
@batrick @tchaikov fixed and rebase. please take a look.

vshankar commented:
minor update w/ spec file changes for the new cephfs header.

vshankar commented:
retest this please

lxbsz commented on src/mds/MetricsHandler.cc:
```diff
@@ -0,0 +1,241 @@
+  std::scoped_lock locker(lock);
+
+  Session *session = mds->get_session(m);
+  dout(20) << __func__ << ": session=" << session << dendl;
```
We could remove all the __func__ , it is duplicated.

```
out/mds.c.log:81256:2019-12-19T06:19:23.530-0500 7fabf99d0700 20 handle_client_metrics: mds.metricshandle_client_metrics: session=0x560d23892880
out/mds.c.log:81626:2019-12-19T06:19:29.162-0500 7fabf99d0700 20 handle_client_metrics: mds.metricshandle_client_metrics: session=0x560d23892880
out/mds.c.log:82074:2019-12-19T06:19:34.282-0500 7fabf99d0700 20 handle_client_metrics: mds.metricshandle_client_metrics: session=0x560d23892880
out/mds.c.log:82462:2019-12-19T06:19:39.401-0500 7fabf99d0700 20 handle_client_metrics: mds.metricshandle_client_metrics: session=0x560d23892880
out/mds.c.log:82951:2019-12-19T06:19:45.545-0500 7fabf99d0700 20 handle_client_metrics: mds.metricshandle_client_metrics: session=0x560d23892880
out/mds.c.log:83327:2019-12-19T06:19:51.178-0500 7fabf99d0700 20 handle_client_metrics: mds.metricshandle_client_metrics: session=0x560d23892880
out/mds.c.log:83804:2019-12-19T06:19:56.297-0500 7fabf99d0700 20 handle_client_metrics: mds.metricshandle_client_metrics: session=0x560d23892880
out/mds.c.log:84137:2019-12-19T06:20:01.418-0500 7fabf99d0700 20 handle_client_metrics: mds.metricshandle_client_metrics: session=0x560d23892880
out/mds.c.log:84650:2019-12-19T06:20:07.567-0500 7fabf99d0700 20 handle_client_metrics: mds.metricshandle_client_metrics: session=0x560d23892880
out/mds.c.log:84992:2019-12-19T06:20:13.193-0500 7fabf99d0700 20 handle_client_metrics: mds.metricshandle_client_metrics: session=0x560d23892880
out/mds.c.log:85466:2019-12-19T06:20:18.314-0500 7fabf99d0700 20 handle_client_metrics: mds.metricshandle_client_metrics: session=0x560d23892880
out/mds.c.log:85821:2019-12-19T06:20:23.432-0500 7fabf99d0700 20 handle_client_metrics: mds.metricshandle_client_metrics: session=0x560d23892880

```

vshankar replied:
ack -- I'll do the changes.

vshankar replied:
`__func__` is duplicated in couple of places in this source. cleanup up that.

vshankar replied:
done

lxbsz commented on src/mds/MetricsHandler.cc:
```diff
@@ -0,0 +1,241 @@
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_mds
+#undef dout_prefix
+#define dout_prefix *_dout << __func__ << ": mds.metrics"
```
Here we may need to add a space at the end, or it will be looks like:

`out/mds.c.log:81256:2019-12-19T06:19:23.530-0500 7fabf99d0700 20 handle_client_metrics: mds.metricshandle_client_metrics: session=0x560d23892880`

lxbsz commented on src/include/cephfs/metrics/Types.h:
```diff
@@ -0,0 +1,284 @@
+
+    template <typename ClientMetricPayload>
+    inline void operator()(ClientMetricPayload &payload) const {
+      using ceph::decode;
```
maybe this could be removed.

vshankar replied:
Thanks for bringing this up -- this is done is every other encode/decode routine in src/include (and elsewhere too). We should clean all this up in a separate PR.

lxbsz replied:
Yeah, make sense.

lxbsz commented:
Test this with the kclient patches, it works:

The kmesg logs:

```
<7>[  867.674007] ceph:  mdsc delayed_work
[...]
<7>[  867.674125] ceph:  cap metric type 0, hit 19, mis 1, total 4
<7>[  867.674127] ceph:  dentry lease metric type 4, hit 1, mis 0, total 3
<7>[  867.674128] ceph:  send_metric mds0 0000000054987168
[...]
```

The ceph perf fs stats:

```
[root@fedora1 build]# ./bin/ceph fs perf stats | python -m json.tool
{
    "client_metadata": {
        "client.4280": {
            "IP": "v1:192.168.195.165",
            "hostname": "fedora1",
            "mount_point": "N/A",
            "root": "////////////dir1//////////dir2"
        }
    },
    "counters": [
        "cap_hit"
    ],
    "global_counters": [
        "read_latency",
        "write_latency",
        "metadata_latency"
    ],
    "global_metrics": {
        "client.4280": [
            [
                0,
                0
            ],
            [
                0,
                0
            ],
            [
                0,
                0
            ]
        ]
    },
    "metrics": {
        "delayed_ranks": [],
        "mds.0": {
            "client.4280": [
                [
                    19,
                    1
                ]
            ]
        }
    }
}
```

Currently the kclient will send the metric per 5 seconds with the existing ping delay_work. Or should it must be a per second ? Won't per second be too frequency ? 

The https://github.com/ceph/ceph/pull/32120/files is sending the perf metric per second.

@vshankar @batrick @jtlayton 
Any idea ?

I am also thinking to add one sysfs to allow kclient user to enable/disable the perf metric sending if we choose per second.

Thanks.

lxbsz commented:
> Test this with the kclient patches, it works:
> 
> The kmesg logs:
> 
> ```
> <7>[  867.674007] ceph:  mdsc delayed_work
> [...]
> <7>[  867.674125] ceph:  cap metric type 0, hit 19, mis 1, total 4
> <7>[  867.674127] ceph:  dentry lease metric type 4, hit 1, mis 0, total 3
> <7>[  867.674128] ceph:  send_metric mds0 0000000054987168
> [...]
> ```
> 

Currently only the cap metric and dentry lease metric are support in the kclient, I will add the dentry lease relating code after this PR merged in MDS side and others if needed.

vshankar commented:
> Currently the kclient will send the metric per 5 seconds with the existing ping delay_work. Or should it must be a per second ? Won't per second be too frequency ?
> 
> The https://github.com/ceph/ceph/pull/32120/files is sending the perf metric per second.

That's because sending metric updates is tied to `Client::tick()`. However, sending every second looks ok to me.

lxbsz commented:
> > Currently the kclient will send the metric per 5 seconds with the existing ping delay_work. Or should it must be a per second ? Won't per second be too frequency ?
> > The https://github.com/ceph/ceph/pull/32120/files is sending the perf metric per second.
> 
> That's because sending metric updates is tied to `Client::tick()`. However, sending every second looks ok to me.

Yeah, already implemented it via every second.
The patch series is: https://patchwork.kernel.org/project/ceph-devel/list/?series=221131

Thanks.

vshankar commented:
@batrick please take a look

vshankar commented:
> From: /ceph/teuthology-archive/pdonnell-2020-01-09_04:58:50-fs-wip-pdonnell-testing-20200108.153813-distro-basic-smithi/4648429/remote/smithi061/log/ceph-mds.c.log.gz

will take a look later today

vshankar commented:
> From: /ceph/teuthology-archive/pdonnell-2020-01-09_04:58:50-fs-wip-pdonnell-testing-20200108.153813-distro-basic-smithi/4648429/remote/smithi061/log/ceph-mds.c.log.gz

this caught a bug in metrics handling (`MetricsHandler` class) which is handled for the most part but there is a small racy window where it still happens. The original issue is when a session is added and removed (say, on rank 1) before and update is sent to rank 0, the next metric update message to rank 0 (from rank 1) would be for removing (untracking) the tracked session, but rank 0 never saw (and tracked) this session update from rank 1. I had handled this case by introducing `MetricsHandler::last_updated_seq` version variable which is incremented when metrics update message is sent to rank 0. That had a bug -- I'll fix and update.

vshankar commented:
@batrick fixed and updated. commit d9ee8265146a2da1d40305c008e722033a550360 fixed how `last_updated_seq` is used.

tchaikov commented on src/mgr/MDSPerfMetricTypes.h:
```diff
@@ -0,0 +1,333 @@
+        it = std::prev(counters->end());
+      }
+      update_counter(descriptor, &(*it));
+      it++;
```
`++it`

tchaikov commented on src/mgr/MDSPerfMetricTypes.h:
```diff
@@ -0,0 +1,333 @@
+      // TODO: optimize
+      if (it == counters->end()) {
+        counters->push_back(PerformanceCounter());
+        it = std::prev(counters->end());
```
```c++
std::advance(it, -1);
```

vshankar replied:
probably when we templatize this (and the respective OSD class), all this would be cleaned up.

tchaikov replied:
sorry, i was wrong. as `push_back()` changes the size of `counters`, `counters` could be reallocated, so `it` is not guaranteed to be a validate iterator after the vector is resized.

tchaikov commented on src/mgr/MDSPerfMetricTypes.cc:
```diff
@@ -0,0 +1,93 @@
+#include <ostream>
+#include "mgr/MDSPerfMetricTypes.h"
+
+std::ostream& operator<<(std::ostream& os, const MDSPerfMetricSubKeyDescriptor &d) {
```
your are replicating `OSDPerfMetricTypes.h` and `OSDPerfMetricTypes.cc`. is it possible to refactor `OSDPerfMetricTypes.*` instead, so we can reuse it?

vshankar replied:
I tried that earlier -- templating the respective class, but couldn't get it fully working (probably needed more work related to templates). I can do that as a follow-up PR.

batrick commented:
In 3h: http://pulpito.ceph.com/?branch=wip-pdonnell-testing-20200115.043629

batrick commented:
@vshankar please rebase.

vshankar commented:
> @vshankar please rebase.

ack

vshankar commented:
@batrick rebased and updated

batrick commented on src/mds/MetricAggregator.cc:
```diff
@@ -0,0 +1,340 @@
+
+  for (auto &rank : active) {
+    dout(20) << ": pinging rank=" << rank << dendl;
+    mds_pinger.send_ping(rank);
```
This races with updates to the MDSMap:

````
2020-01-23T09:06:40.110 INFO:tasks.ceph.mds.a.smithi120.stderr:terminate called after throwing an instance of 'std::out_of_range'
2020-01-23T09:06:40.110 INFO:tasks.ceph.mds.a.smithi120.stderr:  what():  map::at
2020-01-23T09:06:40.110 INFO:tasks.ceph.mds.a.smithi120.stderr:*** Caught signal (Aborted) **
2020-01-23T09:06:40.110 INFO:tasks.ceph.mds.a.smithi120.stderr: in thread 7f59411a8700 thread_name:ms_dispatch
2020-01-23T09:06:40.110 INFO:tasks.ceph.mds.a.smithi120.stderr: ceph version 15.0.0-9652-g0636e28 (0636e283d185bb5d812aee2ce73a767f34213264) octopus (dev)
2020-01-23T09:06:40.110 INFO:tasks.ceph.mds.a.smithi120.stderr: 1: (()+0x12dd0) [0x7f594f933dd0]
2020-01-23T09:06:40.111 INFO:tasks.ceph.mds.a.smithi120.stderr: 2: (gsignal()+0x10f) [0x7f594e3f799f]
2020-01-23T09:06:40.111 INFO:tasks.ceph.mds.a.smithi120.stderr: 3: (abort()+0x127) [0x7f594e3e1cf5]
2020-01-23T09:06:40.111 INFO:tasks.ceph.mds.a.smithi120.stderr: 4: (()+0x8ff8b) [0x7f594edacf8b]
2020-01-23T09:06:40.111 INFO:tasks.ceph.mds.a.smithi120.stderr: 5: (()+0x963ec) [0x7f594edb33ec]
2020-01-23T09:06:40.111 INFO:tasks.ceph.mds.a.smithi120.stderr: 6: (()+0x96447) [0x7f594edb3447]
2020-01-23T09:06:40.111 INFO:tasks.ceph.mds.a.smithi120.stderr: 7: (()+0x966a8) [0x7f594edb36a8]
2020-01-23T09:06:40.111 INFO:tasks.ceph.mds.a.smithi120.stderr: 8: (()+0x91f2f) [0x7f594edaef2f]
2020-01-23T09:06:40.112 INFO:tasks.ceph.mds.a.smithi120.stderr: 9: (MDSRank::send_message_mds(boost::intrusive_ptr<Message> const&, int)+0x74e) [0x564c6327b8de]
2020-01-23T09:06:40.113 INFO:tasks.ceph.mds.a.smithi120.stderr: 10: (MDSPinger::send_ping(int)+0x46d) [0x564c635cf95d]
2020-01-23T09:06:40.113 INFO:tasks.ceph.mds.a.smithi120.stderr: 11: (MetricAggregator::ping_all_active_ranks()+0x195) [0x564c63560a25]
2020-01-23T09:06:40.113 INFO:tasks.ceph.mds.a.smithi120.stderr: 12: (()+0x447c54) [0x564c63560c54]
2020-01-23T09:06:40.113 INFO:tasks.ceph.mds.a.smithi120.stderr: 13: (()+0xc2a33) [0x7f594eddfa33]
2020-01-23T09:06:40.113 INFO:tasks.ceph.mds.a.smithi120.stderr: 14: (()+0x82de) [0x7f594f9292de]
2020-01-23T09:06:40.113 INFO:tasks.ceph.mds.a.smithi120.stderr: 15: (clone()+0x43) [0x7f594e4bc4b3]
````

From: /ceph/teuthology-archive/pdonnell-2020-01-23_00:39:49-kcephfs-wip-pdonnell-testing-20200122.210533-distro-basic-smithi/4695282/teuthology.log

I realize you considered this in the design but I'm afraid that this separate lock won't work. The call to `mds->send_message_mds` relies on `mds.mdsmap` which may be updated asynchronously.

This pinger thread needs to use the `mds_lock`. Or, alternatively, you can make a helper `MDSRank::send_message_mds` which is static and takes an MDSMap. Then store a copy of the MDSMap in `MetricAggregator` for passing to that helper. (I think that should work?)

vshankar replied:
ack -- calls to `send_message_mds` can be wrapped around `mds_lock` -- in `MDSPinger::send-ping` and `MetricsHandler::update_rank0`. I think that should be ok -- the bulk of metrics processing would be outside `mds_lock`.

vshankar replied:
> ack -- calls to `send_message_mds` can be wrapped around `mds_lock` -- in `MDSPinger::send-ping` and `MetricsHandler::update_rank0`. I think that should be ok -- the bulk of metrics processing would be outside `mds_lock`.

This will deadlock -- `MetricAggregator::notify_mdsmap` call lock order would be mds_lock -> locak_lock, but the pinger thread will ping active MDSs with local_lock -> MDSPinger::lock -> mds_lock (to keep minimum critical section under mds_lock). We could probably drop local_lock once a copy of the active set is made, but I think we are better off with maintaining a copy of mds map.

vshankar replied:
> This pinger thread needs to use the `mds_lock`. Or, alternatively, you can make a helper `MDSRank::send_message_mds` which is static and takes an MDSMap. Then store a copy of the MDSMap in `MetricAggregator` for passing to that helper. (I think that should work?)

There *might* be other issues with the static `send_message_mds` approach -- currently `send_message_mds` additionally sends an updated MDSMap to its peer based on the cached epoch of the peer MDS (vs epoch from its mds map). How would the custom `send_message_mds` deal with this?

Right now, sequence numbers are sent from rank 0 to other active ranks and metrics are sent from all active ranks to rank 0 (carrying the sequence number). Ignoring to send updated mds map *might* just cause other problems.

vshankar replied:
@batrick -- I think dropping local_lock before acquiring mds_lock would work. I'll give it a try tomorrow. Thoughts?

batrick replied:
Should work I think.

batrick replied:
> @batrick -- I think dropping local_lock before acquiring mds_lock would work. I'll give it a try tomorrow. Thoughts?

ERr, if you're going to pick up the mds_lock then what's the point of storing the set of active ranks?

vshankar replied:
true. that will be a quirk in the design which has to be lived with until ->send_message_mds is decoupled from MDSRank (w/o the need to acquire mds_lock).

vshankar commented:
@batrick -- fixed and updated. local lock is dropped before acquiring mds_lock and calling ->send_message_mds().

vshankar commented:
updated w/ a minor change.

vshankar commented:
http://pulpito.ceph.com/vshankar-2020-01-29_08:33:18-multimds-wip-inter-mds-stats-distro-basic-smithi/

vshankar commented:
> http://pulpito.ceph.com/vshankar-2020-01-29_08:33:18-multimds-wip-inter-mds-stats-distro-basic-smithi/

@batrick run finished -- please take a look.

vshankar commented:
@batrick ping

batrick commented:
Looks normalish. I am cleaning up some of the test failures and will do a more extensive test run tomorrow.

batrick commented:
jenkins test make check

batrick commented:
http://pulpito.ceph.com/pdonnell-2020-02-14_04:02:50-fs-wip-pdonnell-testing-20200214.001201-distro-basic-smithi/

batrick commented:
`"2020-02-14T04:39:04.420920+0000 mgr.x (mgr.4100) 1 : cluster [ERR] Failed to load ceph-mgr modules: rook" in cluster log `

from http://pulpito.ceph.com/pdonnell-2020-02-14_04:02:50-fs-wip-pdonnell-testing-20200214.001201-distro-basic-smithi/4761526/

is new. @vshankar might be a new problem in master. Can you check? Otherwise I'll have a look tomorrow.

vshankar commented:
> `"2020-02-14T04:39:04.420920+0000 mgr.x (mgr.4100) 1 : cluster [ERR] Failed to load ceph-mgr modules: rook" in cluster log `
> 
> from http://pulpito.ceph.com/pdonnell-2020-02-14_04:02:50-fs-wip-pdonnell-testing-20200214.001201-distro-basic-smithi/4761526/
> 
> is new. @vshankar might be a new problem in master. Can you check? Otherwise I'll have a look tomorrow.

from teuthology log:

```
0-02-14T04:39:03.507 INFO:tasks.ceph.mgr.y.smithi074.stderr:2020-02-14T04:39:03.509+0000 7f384077d040 -1 mgr[py] Module not found: 'rook'
2020-02-14T04:39:03.508 INFO:tasks.ceph.mgr.y.smithi074.stderr:2020-02-14T04:39:03.509+0000 7f384077d040 -1 mgr[py] Traceback (most recent call last):
2020-02-14T04:39:03.508 INFO:tasks.ceph.mgr.y.smithi074.stderr:  File "/usr/share/ceph/mgr/rook/__init__.py", line 2, in <module>
2020-02-14T04:39:03.508 INFO:tasks.ceph.mgr.y.smithi074.stderr:    from .module import RookOrchestrator
2020-02-14T04:39:03.508 INFO:tasks.ceph.mgr.y.smithi074.stderr:  File "/usr/share/ceph/mgr/rook/module.py", line 33, in <module>
2020-02-14T04:39:03.509 INFO:tasks.ceph.mgr.y.smithi074.stderr:    from .rook_cluster import RookCluster
2020-02-14T04:39:03.509 INFO:tasks.ceph.mgr.y.smithi074.stderr:  File "/usr/share/ceph/mgr/rook/rook_cluster.py", line 14, in <module>
2020-02-14T04:39:03.509 INFO:tasks.ceph.mgr.y.smithi074.stderr:    import jsonpatch
2020-02-14T04:39:03.509 INFO:tasks.ceph.mgr.y.smithi074.stderr:ModuleNotFoundError: No module named 'jsonpatch'
```

vshankar commented:
> is new. @vshankar might be a new problem in master. Can you check? Otherwise I'll have a look tomorrow.

commit 846761ef7afab43144f38bf5631fd859d6964820

sebastian-philipp commented:
I think I caused the rook problem. Will have a look

sebastian-philipp commented:
https://tracker.ceph.com/issues/44138

batrick commented on src/mds/MetricAggregator.cc:
```diff
@@ -0,0 +1,342 @@
+  }
+
+  if (pinger.joinable()) {
+    pinger.join();
```
`pinger` needs to acquire `mds_lock` in `update_rank0` causing deadlock here. See

````
2020-02-15T03:18:43.348+0000 16206700 10 shutdown: mds.metrics
2020-02-15T03:18:43.531+0000 1ea16700 20 update_rank0: mds.metrics
2020-02-15T03:18:43.531+0000 1ea16700 20 update_rank0: mds.metrics: sending metric updates for 0 clients to rank 0 with sequence number 348, last updated sequence number 862
2020-02-15T03:18:44.156+0000 1f217700 10 mds.metric.aggregator ping_all_active_ranks: pinging 1 active mds(s)
2020-02-15T03:18:44.156+0000 1f217700 20 mds.metric.aggregator ping_all_active_ranks: pinging rank=0
2020-02-15T03:18:44.156+0000 1f217700 10 mds.pinger send_ping: rank=0
2020-02-15T03:18:44.156+0000 1f217700 10 mds.pinger send_ping: sending ping with sequence=349 to rank=0
````

From: /ceph/teuthology-archive/pdonnell-2020-02-15_01:51:13-fs-wip-pdonnell-testing-20200214.224251-distro-basic-smithi/4764868/remote/smithi177/log/ceph-mds.b.log.gz

vshankar replied:
crap -- lock ordering in shutdown sequence.

vshankar replied:
`MetricAggregator::shutdown()` and `MetricsHandler::shutdown()` would need to drop `mds_lock` before calling `Thread::join()`.

@batrick -- dropping the lock seems to be ok -- `MDSDaemon::suicide()` and `MDSRankDispatcher::shutdown()` sets `stopping` flag as a guard. Thoughts?

batrick replied:
You can use `std::scoped_lock` and `try_lock`. Check stopping flag in `update_rank0` if the lock fails.

Or a `std::queue<std::pair<Message, rank_t>>` of a messages to drain in some tick thread in `MDSRank` holding the `mds_lock`.

batrick replied:
(I don't recommend dropping the `mds_lock` in shutdown.)

vshankar replied:
> (I don't recommend dropping the `mds_lock` in shutdown.)

Any particular reason for that? MDLog::shutdown() drops mds_lock and waits for its thread to terminate.

vshankar replied:
(FWIW, MDCache does the same drop and reacquire mds_lock trickery)

batrick replied:
Hmm, okay you're right. Dropping the lock in this case seems acceptable.

vshankar commented:
@batrick -- the drop mds lock during shutdown is a separate commit (b2fe51d) for now. I'll squash that when this is ready for merge.

gregsfortytwo commented on src/mds/MetricAggregator.cc:
```diff
@@ -53,7 +55,9 @@ int MetricAggregator::init() {
   dout(10) << dendl;
+  ceph_assert(ceph_mutex_is_locked_by_me(mds->mds_lock));
 
+  mds->mds_lock.unlock();
```
Rather than do all this with the real mds_lock, why not just read out the address of mds 0 in notify_mdsmap and send the message using a non-rank-based interface? That seems a lot cleaner with the layers and makes locking simpler.

vshankar replied:
IIRC, Patrick and I discussed somewhat this a while back where a message could be sent to an mds without acquiring mds_lock. I think the point where we decided to go ahead with the current implementation was somewhat related to sending mdsmap before sending out the actual message. Thinking again on this -- does it really matter to send an (updated) mdsmap in this scenario?

gregsfortytwo replied:
Ohhhh is it trying to make sure the receiving MDS always has a current MDSMap before dealing with requests?
Hmm, that is a thing we don't want to break in the general case, but it looks like these messages already handle being delayed or misdirected. You'd have to make sure it doesn't erroneously start thinking it owns a metrics gathering server, and that if an MDS sends an update to the old mds0 it correctly sends all the same updates to the new one — but since these area all "total" rather than "since last update" values I think it should be fine?

vshankar replied:
right -- guess it doesn't matter if we sent some updates to an old active mds. this would simplify the locking bits.

vshankar commented on src/messages/MMDSMetrics.h:
```diff
@@ -0,0 +1,52 @@
+  }
+
+  void print(ostream &out) const override {
+    out << "mds_metrics";
```
ok -- makes sense to add some info about the payload.

Regarding the absence of seq numbers, this does not have to be strict in terms of message ordering, so it ok to be lossy.

vshankar replied:
btw, just fyi, `metrics_message_t` (which is a member variable in this message class) has a sequence number which is used to identify laginness, not for ordering as such.

vshankar commented on src/include/cephfs/metrics/Types.h:
```diff
@@ -0,0 +1,284 @@
+    inline void operator()(const ClientMetricPayload &payload) const {
+      ClientMetricType metric_type = ClientMetricPayload::METRIC_TYPE;
+      m_formatter->dump_string("client_metric_type", stringify(metric_type));
+      payload.dump(m_formatter);
```
maybe -- I'll take a closer look.

vshankar commented on src/mds/MetricsHandler.h:
```diff
@@ -0,0 +1,101 @@
+  // deadlock
+  ceph::spinlock lock;
+
+  // ISN sent by rank0 pinger is 1
```
MDSPinger sens a ping message with initial sequence number as 1.

vshankar commented on src/mds/MetricsHandler.h:
```diff
@@ -0,0 +1,101 @@
+  // locally to figure out if a session got added and removed
+  // within an update to rank 0. note that, this is initialized
+  // with the incoming sequence number sent by rank 0.
+  version_t last_updated_seq = 0;
```
you are right the comment -- it's a bit outdated. `last_updated_seq` is used to determine if a message carrying session information was sent to rank 0. If a session got added and removed before a timer task refreshes session updated/metrics to rank 0, the corresponding entry for metrics can be removed locally and need not be sent to rank 0 (for removal).

vshankar commented on src/mds/MDSPinger.cc:
```diff
@@ -0,0 +1,112 @@
+  std::scoped_lock locker(lock);
+  auto it = ping_state_by_rank.find(rank);
+  if (it == ping_state_by_rank.end()) {
+    derr << ": rank=" << rank << " was never sent ping request." << dendl;
```
This handles a corner case of an mds becoming active, then going down before rank 0 can send a ping request.

gregsfortytwo replied:
That doesn't seem like it's worth logging to derr though?

vshankar replied:
ack

vshankar commented:
> I eventually realized when looking at the spinlocks that you probably did that for fast dispatch? Literally using a spinlock is not really what those requirements mean, though — the important part is to not block. This code may be okay in that regard but it will certainly be better with a mutex, and I can tell you that one very important cycle to avoid is doing anything in fast dispatch which can block on sending a message back out. Looks like that’s okay now because the rank0 update function drops the local lock, but be careful of it when updating.

Everything used mutexes initially, but then @batrick suggested to use spinlocks and it made sense to me. I get the point about not blocking in fast dispatch. Mind explaining a bit more as in why mutexes would be better here?

gregsfortytwo commented:
> Everything used mutexes initially, but then @batrick suggested to use spinlocks and it made sense to me. I get the point about not blocking in fast dispatch. Mind explaining a bit more as in why mutexes would be better here?

The kernel knows about mutexes and that informs scheduling decisions; spinlocks are invisible so if you're unlucky you might get descheduled while holding one. Thanks to futex a mutex is also generally going to be about the same speed as a spinlock is; really the only time I know it's appropriate to use a userspace spinlock is when you need to change a few variables atomically.

vshankar commented:
> The kernel knows about mutexes and that informs scheduling decisions; spinlocks are invisible so if you're unlucky you might get descheduled while holding one. Thanks to futex a mutex is also generally going to be about the same speed as a spinlock is; really the only time I know it's appropriate to use a userspace spinlock is when you need to change a few variables atomically.

Given that with the changes to send metrics updates to mds0 via non-rank interface, locking would only be required to track the updates. I think sticking to spinlocks probably makes sense now?

gregsfortytwo commented:
> > The kernel knows about mutexes and that informs scheduling decisions; spinlocks are invisible so if you're unlucky you might get descheduled while holding one. Thanks to futex a mutex is also generally going to be about the same speed as a spinlock is; really the only time I know it's appropriate to use a userspace spinlock is when you need to change a few variables atomically.
> 
> Given that with the changes to send metrics updates to mds0 via non-rank interface, locking would only be required to track the updates. I think sticking to spinlocks probably makes sense now?

What I'm actually worried about here, aside from it not being a good habit to get into, is that I think there are some memory allocations happening underneath those map operations and that's something to avoid with spinlocks anywhere.

vshankar commented:
> What I'm actually worried about here, aside from it not being a good habit to get into, is that I think there are some memory allocations happening underneath those map operations and that's something to avoid with spinlocks anywhere.

fair enough -- I'll do the changes.

vshankar commented:
@jenkins retest this please

vshankar commented:
@batrick @gregsfortytwo please take a look

vshankar commented:
@batrick rebased and updated -- please take a look

batrick commented:
https://shaman.ceph.com/builds/ceph/wip-pdonnell-testing-20200601.182437/539e4ec08992d77c3ecbbcd6249bd94810c0e669/

Tests coming soon...

batrick commented:
http://pulpito.ceph.com/?branch=wip-pdonnell-testing-20200601.182437

batrick commented:
@vshankar  some of the failures look new. Can you check especially the evicted client warning causes.

vshankar commented:
> @vshankar some of the failures look new. Can you check especially the evicted client warning causes.

checking

vshankar commented:
For test: http://pulpito.ceph.com/pdonnell-2020-06-01_22:23:46-fs-wip-pdonnell-testing-20200601.182437-distro-basic-smithi/5110407/

`client.6362` did not reconnect to the MDS.

MDS notes that clients would reconnect:
```
2020-06-01T23:16:55.330+0000 7ffa98c38700 10 mds.0.server apply_blacklist: killed 0
2020-06-01T23:16:55.330+0000 7ffa98c38700  4 mds.0.73 reconnect_start: killed 0 blacklisted sessions (2 blacklist entries, 1)
2020-06-01T23:16:55.330+0000 7ffa98c38700  1 mds.0.server reconnect_clients -- 1 sessions
2020-06-01T23:16:55.330+0000 7ffa98c38700 10 mds.0.sessionmap dump
2020-06-01T23:16:55.330+0000 7ffa98c38700 10 mds.0.sessionmap client.6362 0x5578fc197180 state open completed {11=0x100000001f7} prealloc_inos [0x10000000003~0x1f4,0x100000001f8~0x3e8] delegated_inos [] used_inos []
2020-06-01T23:16:55.330+0000 7ffa98c38700  7 mds.0.tableserver(snaptable) finish_recovery
```
But `client.6362` never reconnects back -- MDS times out for reconnect and evict the client:

```
2020-06-01T23:17:42.339+0000 7ffa9b43d700  5 mds.beacon.a received beacon reply up:reconnect seq 14 rtt 0
2020-06-01T23:17:43.333+0000 7ffa94c30700 10 mds.0.cache cache not ready for trimming
2020-06-01T23:17:43.333+0000 7ffa94c30700 20 mds.0.cache upkeep thread waiting interval 0.999070644s
2020-06-01T23:17:43.880+0000 7ffa97435700  7 mds.0.server reconnect timed out, 1 clients have not reconnected in time
2020-06-01T23:17:43.880+0000 7ffa97435700  1 mds.0.server reconnect gives up on client.6362 172.21.15.205:0/1374455594
2020-06-01T23:17:43.880+0000 7ffa97435700  0 log_channel(cluster) log [WRN] : evicting unresponsive client smithi205:0 (6362), after waiting 48.5506 seconds during MDS startup
2020-06-01T23:17:43.880+0000 7ffa97435700  1 mds.0.73 Evicting (and blacklisting) client session 6362 (172.21.15.205:0/1374455594)
2020-06-01T23:17:43.880+0000 7ffa97435700  0 log_channel(cluster) log [INF] : Evicting (and blacklisting) client session 6362 (172.21.15.205:0/1374455594) 
```

Doesn't look like this PR is causing client evictions.

lxbsz commented:
> For test: http://pulpito.ceph.com/pdonnell-2020-06-01_22:23:46-fs-wip-pdonnell-testing-20200601.182437-distro-basic-smithi/5110407/
> 
> `client.6362` did not reconnect to the MDS.
> 
> MDS notes that clients would reconnect:
> 
> ```
> 2020-06-01T23:16:55.330+0000 7ffa98c38700 10 mds.0.server apply_blacklist: killed 0
> 2020-06-01T23:16:55.330+0000 7ffa98c38700  4 mds.0.73 reconnect_start: killed 0 blacklisted sessions (2 blacklist entries, 1)
> 2020-06-01T23:16:55.330+0000 7ffa98c38700  1 mds.0.server reconnect_clients -- 1 sessions
> 2020-06-01T23:16:55.330+0000 7ffa98c38700 10 mds.0.sessionmap dump
> 2020-06-01T23:16:55.330+0000 7ffa98c38700 10 mds.0.sessionmap client.6362 0x5578fc197180 state open completed {11=0x100000001f7} prealloc_inos [0x10000000003~0x1f4,0x100000001f8~0x3e8] delegated_inos [] used_inos []
> 2020-06-01T23:16:55.330+0000 7ffa98c38700  7 mds.0.tableserver(snaptable) finish_recovery
> ```
> 
> But `client.6362` never reconnects back -- MDS times out for reconnect and evict the client:
> 
> ```
> 2020-06-01T23:17:42.339+0000 7ffa9b43d700  5 mds.beacon.a received beacon reply up:reconnect seq 14 rtt 0
> 2020-06-01T23:17:43.333+0000 7ffa94c30700 10 mds.0.cache cache not ready for trimming
> 2020-06-01T23:17:43.333+0000 7ffa94c30700 20 mds.0.cache upkeep thread waiting interval 0.999070644s
> 2020-06-01T23:17:43.880+0000 7ffa97435700  7 mds.0.server reconnect timed out, 1 clients have not reconnected in time
> 2020-06-01T23:17:43.880+0000 7ffa97435700  1 mds.0.server reconnect gives up on client.6362 172.21.15.205:0/1374455594
> 2020-06-01T23:17:43.880+0000 7ffa97435700  0 log_channel(cluster) log [WRN] : evicting unresponsive client smithi205:0 (6362), after waiting 48.5506 seconds during MDS startup
> 2020-06-01T23:17:43.880+0000 7ffa97435700  1 mds.0.73 Evicting (and blacklisting) client session 6362 (172.21.15.205:0/1374455594)
> 2020-06-01T23:17:43.880+0000 7ffa97435700  0 log_channel(cluster) log [INF] : Evicting (and blacklisting) client session 6362 (172.21.15.205:0/1374455594) 
> ```
> 
> Doesn't look like this PR is causing client evictions.

This is a known issue, which has been fixed in https://tracker.ceph.com/issues/45665.

vshankar commented:
> This is a known issue, which has been fixed in https://tracker.ceph.com/issues/45665.

thx @lxbsz

batrick commented:
https://tracker.ceph.com/issues/45829

otherwise looks normal.
